### DISCUSSION TITLE
search(field, "*") -> noop?

### DISCUSSION THREAD

**Discussion by olibanjoli** [ravendb/ravendb#14170] [General]
We recently had a query that suddenly became very slow. according to traffic watch it took 2000-8000ms. the document collection behind the index contains 56k docs, no fanout index. the docs have a reasonable size, around 7kb. the query included a search() statement. pagesize 50.
the query itself hasn't changed much since it became slow, but it is getting a lot more hits (also concurrently) and we also upgraded the server from 5.3.0 to 5.3.102 since it became slower. the query was fine for 4 years, during that time the collection grew from 0 to 56k docs.

after some analysis i noticed that the query was slow when it contained a ```search(field, '*')``` statement (basically when the user did not enter anything on the search field for the fuzzy search on a table in the ui). 

so the obvious fix for us was to check against that and not to do a ```query.Search(f => f.SomeProperty, "*")``` in our code. problem solved. 

I am wondering why ```search(field, "*")``` does not translate into a noop on the server side as it should match any index entry and not cause any additional work for the server to respond to the query.

i totally understand that we should not add unnecessary statements to the query and rely on the query optimizer. but i would also like to understand more what happens inside the database. i guess my assumption that search(field, "*") should be a noop is wrong but i don't understand why.

kudos to the raven studio traffic watch & the ability to run a query from there and also modify it. i could narrow down the issue of a slow query within minutes to the single statement that caused the problem.

**Comment by ppekrol** [comment]
Hi, we will look into that, I've created an issue to track progress: https://issues.hibernatingrhinos.com/issue/RavenDB-18580

**Comment by shaharhikri** [comment]
Hey,
Can you give us a little more details?
What's the full query you are using?
What's the size of the collection you're trying to run the query on?
An example of the document you're searching in the query will help too.

It's weird because in the second time you run the query, the auto index that related to this search 
should be already exist in the DB, so only the first running of the query should take a little bit longer.

**Reply by ayende** [reply]
To give some context, a query like `search(Name, "*")` is _not_ actually a noop.
Instead, this is actually translated to `exists(Name)`

**Reply by olibanjoli** [reply]
From a user of the c# typed raven client api perspective I would question why it translates into a `exist(Name)`. Because it should always exist. It is defined on the index definition which is used to do the query. 

It used to be a noop for years, why the change now? It also seems very expensive as our queries were very slow on a relatively small collection.

Not sure if the question from @shaharhikri was directed to me or not?

**Reply by ayende** [reply]
Consider the following scenario:

* You have Users with (`Name`, `Email`)
* Your collection have 100,000 items
* You add a property `Title` to your model
* You define an index `from u as docs.Users select new { u.Name, u.Title }`
* You write 10 new users with the `Title` there.

You issue a query on: `Search(Title, '*')`

What would you expect this to return?

You have 100,000 docs without a `Title` field at all. It isn't null, it isn't there.
While you have 10 docs that have `Title` of various values.
